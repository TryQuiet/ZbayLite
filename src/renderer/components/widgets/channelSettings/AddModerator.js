import React from 'react'
import PropTypes from 'prop-types'
import * as R from 'ramda'
import Immutable from 'immutable'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { withStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'
import { AutoSizer } from 'react-virtualized'
import { Scrollbars } from 'react-custom-scrollbars'

import Modal from '../../ui/Modal'
import UserListItem from './UserListItem'

const styles = theme => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  title: {
    marginBottom: 24
  }
})
export const AddModerator = ({
  classes,
  open,
  handleClose,
  members,
  users,
  addModerator
}) => {
  const channelsArray = Array.from(members)
  const options = channelsArray.map(member => {
    if (users.member) {
      return users.member.nickname
    } else {
      return member
    }
  })
  const [input, setInput] = React.useState('')
  return (
    <Modal
      open={open}
      handleClose={() => {
        handleClose()
      }}
      title=''
      fullPage
    >
      <Grid className={classes.root}>
        <Grid
          container
          justify='flex-start'
          direction='column'
          className={classes.fullContainer}
        >
          <Typography variant='h3' className={classes.title}>
            Add a moderator
          </Typography>

          <TextField
            id='outlined-helperText'
            placeholder='Search for usernames'
            variant='outlined'
            onChange={e => setInput(e.target.value)}
          />
          <AutoSizer>
            {({ width, height }) => (
              <Scrollbars
                autoHideTimeout={500}
                style={{ width: width, height: height, overflowX: 'hidden' }}
              >
                {options
                  .filter(nickname => nickname ? nickname.startsWith(input || '') : false)
                  .map(nickname => {
                    return (
                      <UserListItem
                        key={nickname}
                        disableConfirmation
                        name={nickname.substring(0, 20)}
                        actionName='Add a moderator'
                        action={() => {
                          if (
                            users.findKey(user => user.nickname === nickname)
                          ) {
                            addModerator(
                              users.findKey(user => user.nickname === nickname)
                            )
                          } else {
                            addModerator(nickname)
                          }
                          handleClose()
                        }}
                      />
                    )
                  })}
              </Scrollbars>
            )}
          </AutoSizer>
        </Grid>
      </Grid>
    </Modal>
  )
}

AddModerator.propTypes = {
  classes: PropTypes.object.isRequired,
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  addModerator: PropTypes.func.isRequired,
  members: PropTypes.instanceOf(Set).isRequired,
  users: PropTypes.instanceOf(Immutable.Map).isRequired
}
export default R.compose(React.memo, withStyles(styles))(AddModerator)
