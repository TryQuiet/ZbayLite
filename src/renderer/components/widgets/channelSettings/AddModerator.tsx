import React from 'react'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'
import { AutoSizer } from 'react-virtualized'
import { Scrollbars } from 'rc-scrollbars'

import Modal from '../../ui/Modal/Modal'
import UserListItem from './UserListItem'

const useStyles = makeStyles(() => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  title: {
    marginBottom: 24
  }
}))

interface AddModeratorProps {
  open: boolean
  handleClose: () => void
  members: any[]
  users: {
    member: {
      nickname: string
    }
  }
  addModerator: (name: string) => void
}

export const AddModerator: React.FC<AddModeratorProps> = ({
  open,
  handleClose,
  members,
  users,
  addModerator
}) => {
  const classes = useStyles({})
  const channelsArray = Array.from(members)
  const options = channelsArray.map(member => {
    if (users.member) {
      return users.member.nickname
    } else {
      return member
    }
  })
  const [input, setInput] = React.useState('')
  return (
    <Modal
      open={open}
      handleClose={() => {
        handleClose()
      }}
      title=''
      fullPage
    >
      <Grid className={classes.root}>
        <Grid
          container
          justify='flex-start'
          direction='column'
          className={classes.fullContainer}
        >
          <Typography variant='h3' className={classes.title}>
            Add a moderator
          </Typography>

          <TextField
            id='outlined-helperText'
            placeholder='Search for usernames'
            variant='outlined'
            onChange={e => setInput(e.target.value)}
          />
          <AutoSizer>
            {({ width, height }) => (
              <Scrollbars
                autoHideTimeout={500}
                style={{ width: width, height: height, overflowX: 'hidden' }}
              >
                {options
                  .filter(nickname => nickname ? nickname.startsWith(input || '') : false)
                  .map(nickname => {
                    return (
                      <UserListItem
                        key={nickname}
                        disableConfirmation
                        name={nickname.substring(0, 20)}
                        actionName='Add a moderator'
                        action={() => {
                          if (
                            Object.keys(users).find(key => users[key] === nickname)
                          ) {
                            addModerator(
                              Object.keys(users).find(key => users[key] === nickname)
                            )
                          } else {
                            addModerator(nickname)
                          }
                          handleClose()
                        }}
                      />
                    )
                  })}
              </Scrollbars>
            )}
          </AutoSizer>
        </Grid>
      </Grid>
    </Modal>
  )
}

export default AddModerator
