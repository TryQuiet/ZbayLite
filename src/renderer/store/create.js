import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
