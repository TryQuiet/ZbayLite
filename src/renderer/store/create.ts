import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
