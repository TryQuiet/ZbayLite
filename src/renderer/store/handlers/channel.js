import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'
import crypto from 'crypto'
import * as R from 'ramda'
import { remote } from 'electron'
import { DateTime } from 'luxon'

import history from '../../../shared/history'
import operationsHandlers from './operations'
import notificationsHandlers from './notifications'
// import messagesQueueHandlers from './messagesQueue'
import messagesHandlers, { _checkMessageSize } from './messages'
import channelsHandlers from './channels'
import offersHandlers from './offers'
import channelSelectors from '../selectors/channel'
import identitySelectors from '../selectors/identity'
import contactsSelectors from '../selectors/contacts'
import client from '../../zcash'
import { messages } from '../../zbay'
import { errorNotification, LoaderState } from './utils'
import { messageType, actionTypes } from '../../../shared/static'
import { DisplayableMessage } from '../../zbay/messages'
import usersSelectors from '../selectors/users'
import contactsHandlers from './contacts'
import electronStore from '../../../shared/electronStore'
import { channelToUri } from '../../../renderer/zbay/channels'

export const ChannelState = Immutable.Record(
  {
    spentFilterValue: new BigNumber(0),
    id: null,
    message: Immutable.Map(),
    shareableUri: '',
    address: '',
    loader: LoaderState({ loading: false }),
    members: null,
    showInfoMsg: true,
    isSizeCheckingInProgress: false,
    messageSizeStatus: null,
    displayableMessageLimit: 50
  },
  'ChannelState'
)

export const initialState = ChannelState()

const setSpentFilterValue = createAction(
  actionTypes.SET_SPENT_FILTER_VALUE,
  (_, value) => value
)
const setMessage = createAction(actionTypes.SET_CHANNEL_MESSAGE)
const setDisplayableLimit = createAction(actionTypes.SET_DISPLAYABLE_LIMIT)
const setChannelId = createAction(actionTypes.SET_CHANNEL_ID)
const setLoading = createAction(actionTypes.SET_CHANNEL_LOADING)
const setLoadingMessage = createAction(actionTypes.SET_CHANNEL_LOADING_MESSAGE)
const setShareableUri = createAction(actionTypes.SET_CHANNEL_SHAREABLE_URI)
const setAddress = createAction(actionTypes.SET_CHANNEL_ADDRESS)
const resetChannel = createAction(actionTypes.SET_CHANNEL)
const isSizeCheckingInProgress = createAction(
  actionTypes.IS_SIZE_CHECKING_IN_PROGRESS
)
const messageSizeStatus = createAction(actionTypes.MESSAGE_SIZE_STATUS)

export const actions = {
  setLoading,
  setLoadingMessage,
  setSpentFilterValue,
  setMessage,
  setShareableUri,
  setChannelId,
  resetChannel,
  isSizeCheckingInProgress,
  messageSizeStatus,
  setDisplayableLimit
}

const loadChannel = key => async (dispatch, getState) => {
  try {
    dispatch(setChannelId(key))
    dispatch(setDisplayableLimit(30))
    // Calculate URI on load, that way it won't be outdated, even if someone decides
    // to update channel in vault manually
    const contact = contactsSelectors.contact(key)(getState())
    const unread = contact.newMessages.size
    remote.app.badgeCount = remote.app.badgeCount - unread
    const ivk = electronStore.get(
      `defaultChannels.${contact.get('address')}.keys.ivk`
    )
    if (ivk) {
      const uri = await channelToUri({
        name: contact.get('username'),
        ivk
      })
      dispatch(setShareableUri(uri))
    }
    electronStore.set(
      `lastSeen.${key}`,
      parseInt(DateTime.utc().toSeconds()).toString()
    )
    dispatch(setAddress(contact.address))

    dispatch(contactsHandlers.actions.cleanNewMessages({ contactAddress: key }))
    // await dispatch(clearNewMessages())
    // await dispatch(updateLastSeen())
  } catch (err) {}
}
const loadOffer = (id, address) => async (dispatch, getState) => {
  try {
    await dispatch(offersHandlers.epics.updateLastSeen({ itemId: id }))
    dispatch(setChannelId(id))
    dispatch(setShareableUri(''))
    dispatch(setAddress(address))
  } catch (err) {}
}
const linkChannelRedirect = targetChannel => async (dispatch, getState) => {
  const contact = contactsSelectors.contact(targetChannel.address)(getState())
  if (contact.address) {
    history.push(`/main/channel/${targetChannel.address}`)
    return
  }

  // We can parse timestamp to blocktime and get accurate birthday block for this channel
  // Skipped since we dont support rescaning also we already got birthday of zbay as main wallet birthday
  await client.importKey(targetChannel.keys.ivk)
  await dispatch(
    contactsHandlers.actions.addContact({
      key: targetChannel.address,
      contactAddress: targetChannel.address,
      username: targetChannel.name
    })
  )
  const importedChannels = electronStore.get(`importedChannels`) || {}
  electronStore.set('importedChannels', {
    ...importedChannels,
    [targetChannel.address]: {
      address: targetChannel.address,
      name: targetChannel.name,
      description: targetChannel.description,
      owner: targetChannel.owner,
      keys: targetChannel.keys
    }
  })
  history.push(`/main/channel/${targetChannel.address}`)
}
const sendOnEnter = (event, resetTab) => async (dispatch, getState) => {
  if (resetTab) {
    resetTab(0)
  }
  const enterPressed = event.nativeEvent.keyCode === 13
  const shiftPressed = event.nativeEvent.shiftKey === true
  const channel = channelSelectors.channel(getState()).toJS()
  const messageToSend = channelSelectors.message(getState())
  let message
  if (enterPressed && !shiftPressed) {
    event.preventDefault()
    const privKey = identitySelectors.signerPrivKey(getState())
    message = messages.createMessage({
      messageData: {
        type: messageType.BASIC,
        data: messageToSend
      },
      privKey: privKey
    })
    const isMergedMessageTooLong = await dispatch(
      _checkMessageSize(message.message)
    )
    if (!isMergedMessageTooLong) {
      dispatch(setMessage(''))
      const myUser = usersSelectors.myUser(getState())
      const messageDigest = crypto.createHash('sha256')

      const messageEssentials = R.pick(['createdAt', 'message', 'spent'])(
        message
      )
      const key = messageDigest
        .update(JSON.stringify(messageEssentials))
        .digest('hex')

      const messagePlaceholder = DisplayableMessage({
        ...message,
        id: key,
        sender: {
          replyTo: myUser.address,
          username: myUser.nickname
        },
        fromYou: true,
        status: 'pending',
        message: messageToSend
      })
      dispatch(
        contactsHandlers.actions.addMessage({
          key: channel.id,
          message: { [key]: messagePlaceholder }
        })
      )
      dispatch(
        operationsHandlers.actions.addOperation({
          channelId: channel.id,
          id: key
        })
      )
      const identityAddress = identitySelectors.address(getState())

      const transfer = await messages.messageToTransfer({
        message: message,
        address: channel.address,
        identityAddress
      })
      const transaction = await client.sendTransaction(transfer)
      console.log(transaction, 'transaction details')
      if (!transaction.txid) {
        dispatch(
          contactsHandlers.actions.addMessage({
            key: channel.id,
            message: { [key]: messagePlaceholder.set('status', 'failed') }
          })
        )
        dispatch(
          notificationsHandlers.actions.enqueueSnackbar(
            errorNotification({
              message:
                "Couldn't send the message, please check node connection."
            })
          )
        )
        return
      }
      dispatch(
        operationsHandlers.epics.resolvePendingOperation({
          channelId: channel.id,
          id: key,
          txid: transaction.txid
        })
      )
    }
  }
}
const sendChannelSettingsMessage = ({
  address,
  minFee = '0',
  onlyRegistered = '0'
}) => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const owner = identitySelectors.signerPubKey(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.CHANNEL_SETTINGS,
      data: {
        owner,
        minFee,
        onlyRegistered
      }
    },
    privKey: privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: address,
    identityAddress
  })
  try {
    const txid = await client.sendTransaction(transfer)
    if (txid.error) {
      throw new Error(txid.error)
    }
    return 1
  } catch (err) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't create channel, please check node connection."
        })
      )
    )
    return -1
  }
}

const resendMessage = messageData => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const channel = channelSelectors.data(getState()).toJS()
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageData.type,
      data: messageData.message,
      spent: parseFloat(messageData.spent.toString())
    },
    privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: channel.address,
    amount: parseFloat(messageData.spent.toString()),
    identityAddress
  })
  const messagePlaceholder = DisplayableMessage({
    ...messageData,
    status: 'pending'
  })
  dispatch(
    contactsHandlers.actions.addMessage({
      key: channel.key,
      message: { [messageData.id]: messagePlaceholder }
    })
  )
  const transaction = await client.sendTransaction(transfer)
  if (!transaction.txid) {
    dispatch(
      contactsHandlers.actions.addMessage({
        key: channel.key,
        message: {
          [messageData.id]: messagePlaceholder.set('status', 'failed')
        }
      })
    )
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't send the message, please check node connection."
        })
      )
    )
    return
  }
  dispatch(
    operationsHandlers.epics.resolvePendingOperation({
      channelId: channel.key,
      id: messageData.id,
      txid: transaction.txid
    })
  )
}

const updateLastSeen = () => async (dispatch, getState) => {
  const channelId = channelSelectors.channelId(getState())
  return dispatch(channelsHandlers.epics.updateLastSeen({ channelId }))
}

const clearNewMessages = () => async (dispatch, getState) => {
  const channelId = channelSelectors.channelId(getState())
  dispatch(messagesHandlers.actions.cleanNewMessages({ channelId }))
}

export const epics = {
  sendOnEnter,
  loadChannel,
  resendMessage,
  clearNewMessages,
  updateLastSeen,
  loadOffer,
  sendChannelSettingsMessage,
  linkChannelRedirect
}

// TODO: we should have a global loader map
export const reducer = handleActions(
  {
    [setLoading]: (state, { payload: loading }) =>
      state.setIn(['loader', 'loading'], loading),
    [setLoadingMessage]: (state, { payload: message }) =>
      state.setIn(['loader', 'message'], message),
    [setSpentFilterValue]: (state, { payload: value }) =>
      state.set('spentFilterValue', new BigNumber(value)),
    [setMessage]: (state, { payload: value }) =>
      state.setIn(['message', state.get('id')], value),
    [setChannelId]: (state, { payload: id }) => state.set('id', id),
    [isSizeCheckingInProgress]: (state, { payload }) =>
      state.set('isSizeCheckingInProgress', payload),
    [messageSizeStatus]: (state, { payload }) =>
      state.set('messageSizeStatus', payload),
    [setShareableUri]: (state, { payload: uri }) =>
      state.set('shareableUri', uri),
    [setDisplayableLimit]: (state, { payload: limit }) =>
      state.set('displayableMessageLimit', limit),
    [setAddress]: (state, { payload: address }) =>
      state.set('address', address),
    [resetChannel]: () => initialState
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
