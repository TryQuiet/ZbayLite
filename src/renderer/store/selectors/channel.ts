import { createSelector } from 'reselect'
import identitySelectors from './identity'
import messagesQueueSelectors from './messagesQueue'
import contacts from './contacts'
import { networkFee, messageType } from '../../../shared/static'
import publicChannels from './publicChannels'
import users from './users'

import { Store } from '../reducers'
import { DisplayableMessage } from '../../zbay/messages.types'

const channel = (s: Store) => s.channel

export const channelInfo = createSelector(channel, ch => {
  const channel = {
    ...ch
  }
  delete channel.message
  return channel
}) // TODO refactor

const isPublicChannel = createSelector(
  publicChannels.publicChannels,
  channel,
  (pubChannels, channel) => {
    if (pubChannels && channel) {
      const publicChannelAddresses = Object.values(pubChannels).map(el => el.address)
      const { address } = channel
      return publicChannelAddresses.includes(address)
    } else {
      return false
    }
  }
)

export const spentFilterValue = createSelector(channel, c =>
  c.spentFilterValue ? spentFilterValue : -1
)

export const message = createSelector(channel, c => c.message[c.id] || '')
export const id = createSelector(channel, c => c.id)
const data = createSelector(contacts.contacts, id, (channels, id) => channels[id])
export const isSizeCheckingInProgress = createSelector(channel, c => c.isSizeCheckingInProgress)
export const messageSizeStatus = createSelector(channel, c => c.messageSizeStatus)
export const displayableMessageLimit = createSelector(channel, c => c.displayableMessageLimit)
export const isOwner = createSelector(
  id,
  contacts.contacts,
  identitySelectors.signerPubKey,
  (id, con, myKey) => {
    const contact = con[id]
    if (!contact) {
      return false
    }
    const settingsMsg = Array.from(Object.values(contact.messages)).filter(
      msg => msg.type === messageType.CHANNEL_SETTINGS
    )[0]
    if (settingsMsg && settingsMsg.message.owner === myKey) {
      return true
    }
    return false
  }
)
export const channelSettingsMessage = createSelector(data, data => {
  if (!data) {
    return null
  }
  const settingsMsg = Array.from(Object.values(data.messages)).filter(
    msg =>
      msg.type === messageType.CHANNEL_SETTINGS || msg.type === messageType.CHANNEL_SETTINGS_UPDATE
  )
  if (!settingsMsg.length) {
    return null
  }
  return settingsMsg.reduce((prev, curr) => (prev.createdAt > curr.createdAt ? prev : curr))
})
export const advertFee = createSelector(channelSettingsMessage, settingsMsg => {
  if (settingsMsg === null) {
    return 0
  }
  return settingsMsg.message.minFee || settingsMsg.message.updateMinFee
})
export const channelDesription = createSelector(channelSettingsMessage, settingsMsg => {
  if (settingsMsg === null) {
    return 0
  }
  return settingsMsg.message.updateChannelDescription || ''
})
export const onlyRegistered = createSelector(channelSettingsMessage, settingsMsg => {
  if (settingsMsg === null) {
    return 0
  }
  return settingsMsg.message.updateOnlyRegistered || '0'
})
export const unread = createSelector(data, data => (data ? data.unread : 0))

export const queuedMessages = createSelector(
  messagesQueueSelectors.queue,
  channel,
  (queue, channel) => queue.filter(m => m.channelId === channel.id)
)

export const loader = createSelector(channel, meta => meta.loader)

const checkMessageTargetTimeWindow = ({ targetCreatedAt, timeStamp, timeWindow }) => {
  const inRange = ({ timeStamp, targetCreatedAt, timeWindow }) => {
    return (timeStamp - targetCreatedAt) * (timeStamp - timeWindow) <= 0
  }
  return inRange({ timeStamp, targetCreatedAt, timeWindow })
}

const concatMessages = (mainMsg, messagesToConcat) => {
  if (messagesToConcat.length === 1) {
    return mainMsg
  } else {
    const messagesArray = messagesToConcat.map(msg => msg.message)
    const lastMessageStatus = messagesToConcat[messagesToConcat.length - 1].status
    const concatedMessages = messagesArray.join('\n')
    const mergedMessage = {
      ...mainMsg,
      message: concatedMessages,
      status: lastMessageStatus
    }
    return mergedMessage
  }
}

export const mergeIntoOne = (messages: DisplayableMessage[]) => {
  if (messages.length === 0) return
  const result = [[]]
  let last: DisplayableMessage = null
  for (const msg of messages) {
    const isMessageInTargetZone = last
      ? checkMessageTargetTimeWindow({
        targetCreatedAt: last.createdAt,
        timeStamp: msg.createdAt,
        timeWindow: last.createdAt + 300
      })
      : true
    if (last && msg.status === 'failed') {
      result.push([])
      result[result.length - 1].push(msg)
    } else if (last && (msg.type !== 1 || last.type !== 1)) {
      result.push([])
      result[result.length - 1].push(msg)
    } else if ((last && last.sender.replyTo !== msg.sender.replyTo) || !isMessageInTargetZone) {
      result.push([])
      result[result.length - 1].push(msg)
    } else {
      result[result.length - 1].push(msg)
    }
    last = msg
  }
  const concatedMessages = result.map(array => {
    return concatMessages(array[0], array)
  })
  return concatedMessages
}

export const shareableUri = createSelector(channel, c => c.shareableUri)

export const inputLocked = createSelector(
  identitySelectors.balance('zec'),
  identitySelectors.lockedBalance('zec'),
  users.users,
  identitySelectors.signerPubKey,
  (available, locked, users, signerPubKey) => {
    if (available.gt(networkFee)) {
      if (users[signerPubKey]) {
        if (users[signerPubKey].createdAt) {
          return INPUT_STATE.AVAILABLE
        } else {
          return INPUT_STATE.UNREGISTERED
        }
      }
    } else {
      if (locked.gt(0)) {
        return INPUT_STATE.LOCKED
      }
    }
    return INPUT_STATE.DISABLE
  }
)

export const INPUT_STATE = {
  DISABLE: 0,
  AVAILABLE: 1,
  LOCKED: 2,
  UNREGISTERED: 3
}

export const channelId = createSelector(channel, ch => ch.id)

export const members = createSelector(contacts.contacts, id, (c, channelId) => {
  const contact = c[channelId]
  if (!contact) {
    return new Set()
  }
  return Array.from(Object.values(contact.messages)).reduce((acc, msg) => {
    return acc.add(msg.sender.replyTo)
  }, new Set())
})

export const channelParticipiants = createSelector(contacts.contacts, id, (c, i) => {
  const contact = c[i]
  if (!contact) {
    return new Set()
  }
  const messages = contact.messages
  const members = Array.from(Object.values(messages)).reduce((acc, msg) => {
    return acc.add(msg.sender.replyTo)
  }, new Set())
  return members
})

export default {
  data,
  inputLocked,
  loader,
  channel,
  spentFilterValue,
  message,
  shareableUri,
  channelId,
  channelInfo,
  advertFee,
  onlyRegistered,
  members,
  channelParticipiants,
  unread,
  messageSizeStatus,
  isSizeCheckingInProgress,
  id,
  isOwner,
  channelDesription,
  displayableMessageLimit,
  isPublicChannel
}
