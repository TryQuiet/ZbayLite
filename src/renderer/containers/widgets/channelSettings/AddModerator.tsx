import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import * as R from 'ramda'

// import moderationActions from '../../../store/handlers/moderationActions'
import AddModerator from '../../../components/widgets/channelSettings/AddModerator'
import channelSelectors from '../../../store/selectors/channel'
import usersSelector from '../../../store/selectors/users'
import { withModal } from '../../../store/handlers/modals'
import contactsSelectors from '../../../store/selectors/contacts'

export const mapStateToProps = state => {
  const messagesState = contactsSelectors.directMessages(channelSelectors.channelId(state))(state)
  const visibleMessages = messagesState.visibleMessages
  // const moderators = messagesState.channelModerators
  if (visibleMessages) {
    // const members = visibleMessages.reduce((acc, msg) => {
    //   return acc.add(msg.publicKey)
    // }, new Set())
    return {
      members: [],
      users: usersSelector.users(state)
    }
  } else {
    return {
      members: [],
      users: usersSelector.users(state)
    }
  }
}

export const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      // addModerator: publicKey =>
      //   moderationActions.epics.handleModerationAction({
      //     moderationType: moderationActionsType.ADD_MOD,
      //     moderationTarget: publicKey
      //   })
    },
    dispatch
  )

export default R.compose(
  connect(mapStateToProps, mapDispatchToProps),
  withModal('addModerator')
)(AddModerator)
