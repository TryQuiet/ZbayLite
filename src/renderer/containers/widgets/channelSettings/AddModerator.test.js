/* eslint import/first: 0 */
import BigNumber from 'bignumber.js'

import { mapDispatchToProps, mapStateToProps } from './AddModerator'

import { ChannelState } from '../../../store/handlers/channel'

import create from '../../../store/create'

const channelId = 'randomid'
const baseStore = {
  channel: {
    ...ChannelState,
    spentFilterValue: 38,
    id: channelId,
    shareableUri: 'testuri',
    members: new BigNumber(0),
    message: 'Message written in the input'
  },
  certificates: {
    usersCertificates: ['MIIBtjCCAVwCBgF6Urb0VDAKBggqhkjOPQQDAjASMRAwDgYDVQQDEwdaYmF5IENBMB4XDTIxMDYyODEzMDIzOVoXDTMwMTIzMTIzMDAwMFowgZcxgZQwFAYKKwYBBAGDjBsCARMGZGFtaWFuMD8GA1UEAxM4YXp2eTcybmV4NXZzeWd2NW1ma3M0c2YyN2pycWRoZXpueDIzYm1xeHE3YzdtajNibW5pc3hhaWQwOwYJKwYBAgEPAwEBEy5RbVpWZnhrNWpQb2tiVmlqODNhWVV4ZXRaeWF5bnZCVlRCaGZQUlBSM3hndVZLMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAErEDIPiIQsZ020N34a+KtVLURm+68lyq0T7eklOlE6G3matt9vLKadwM4pNzJRhRigO1W0MNH3uOrwG1o2UOOX6MdMBswDAYDVR0TBAUwAwIBAzALBgNVHQ8EBAMCAAYwCgYIKoZIzj0EAwIDSAAwRQIgAocVe47k6ok0wA9vFaMRUxiD7syVPh+d3YIKLHanj5UCIQCNiXIh8DKsfkSICqzG3LxWl15W5Rq7wVij1fdM6AXo7Q=='],
    ownCertificate: {
      certificate: 'MIIBtjCCAVwCBgF6Urb0VDAKBggqhkjOPQQDAjASMRAwDgYDVQQDEwdaYmF5IENBMB4XDTIxMDYyODEzMDIzOVoXDTMwMTIzMTIzMDAwMFowgZcxgZQwFAYKKwYBBAGDjBsCARMGZGFtaWFuMD8GA1UEAxM4YXp2eTcybmV4NXZzeWd2NW1ma3M0c2YyN2pycWRoZXpueDIzYm1xeHE3YzdtajNibW5pc3hhaWQwOwYJKwYBAgEPAwEBEy5RbVpWZnhrNWpQb2tiVmlqODNhWVV4ZXRaeWF5bnZCVlRCaGZQUlBSM3hndVZLMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAErEDIPiIQsZ020N34a+KtVLURm+68lyq0T7eklOlE6G3matt9vLKadwM4pNzJRhRigO1W0MNH3uOrwG1o2UOOX6MdMBswDAYDVR0TBAUwAwIBAzALBgNVHQ8EBAMCAAYwCgYIKoZIzj0EAwIDSAAwRQIgAocVe47k6ok0wA9vFaMRUxiD7syVPh+d3YIKLHanj5UCIQCNiXIh8DKsfkSICqzG3LxWl15W5Rq7wVij1fdM6AXo7Q==',
      privateKey: 'MIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgUMBHFeLut2Tyr5T5hkEtnbh4vx1XSEx2fBTwv29FytagCgYIKoZIzj0DAQehRANCAASsQMg+IhCxnTbQ3fhr4q1UtRGb7ryXKrRPt6SU6UTobeZq2328spp3Azik3MlGFGKA7VbQw0fe46vAbWjZQ45f'
    }
  }
}
describe('Send message popover', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      ...baseStore
    })
  })
  it('will receive right props', () => {
    const state = mapStateToProps(store.getState())
    expect(state).toMatchSnapshot()
  })

  it('will receive right actions', () => {
    const actions = mapDispatchToProps(x => x)
    expect(actions).toMatchSnapshot()
  })
})
