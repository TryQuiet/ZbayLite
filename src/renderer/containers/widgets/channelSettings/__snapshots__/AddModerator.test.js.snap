// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send message popover will receive right actions 1`] = `
Object {
  "addModerator": [Function],
}
`;

exports[`Send message popover will receive right props 1`] = `
Object {
  "members": Set {},
  "users": Object {},
}
`;
